### PR TITLE
[draft] io: consume budget in Registration::readiness

### LINKED ISSUE #4782: Tokio in current_thread not releasing open file handles once the limit is reached.
**Version**

`cargo tree | grep tokio`
```
└── tokio v1.19.2
    └── tokio-macros v1.8.0 (proc-macro)
```
**Platform**
The output of `uname -a` (UNIX), or version and 32 or 64-bit (Windows)
```
Linux pdebian 5.10.0-10-amd64 #1 SMP Debian 5.10.84-1 (2021-12-08) x86_64 GNU/Linux
```

**Description**

While reading from a Unix Socket over a buffered stream, the tokio executor in current_thread doesn't release open files once the open file OS limit is hit and will not accept any more incoming requests, but if the number of open files stays within the OS limit, the file handles are released.

If the executor is switched to a multi_threaded one, this issue doesn't happen. Although the open file limits are hit, it frees the openfiles and the program continues to work.

I am not sure if I am using the BufStream wrong or I might have overlooked something.

I tried this code:

## A minimal version of the server code that accepts connection over a unix socket and prints it

`src/main.rs`
```rust
use anyhow::{Context, Result};
use clap::{App, Arg};
use env_logger::{Builder, Env};
use log::{error, info};
use tokio::{
    io::{AsyncBufReadExt, AsyncWriteExt, BufStream},
    net::{UnixListener, UnixStream},
    runtime,
    // time::{timeout, Duration},
};

pub static SOCK_PATH: &str = "/var/run/sock.sock";

fn main() -> Result<()> {
    Builder::from_env(Env::default().default_filter_or("info")).init();
    let clap_app = App::new("openfiles")
        .version(env!("CARGO_PKG_VERSION"))
        .author(env!("CARGO_PKG_AUTHORS"))
        .about("checking tokio openfiles")
        .arg(
            Arg::new("worker-threads")
                .long("worker-threads")
                .takes_value(true)
                .help("number of worker threads. 0 = current_thread. >0 = worker_threads")
                .default_value("1")
                .global(true),
        )
        .get_matches();
    let threads = clap_app
        .value_of("worker-threads")
        .unwrap()
        .parse::<usize>()
        .unwrap();
    let rt = match threads {
        0 => {
            info!("running in current thread");
            runtime::Builder::new_current_thread()
                .enable_all()
                .build()
                .context("cannot create runtime")?
        }
        multi => {
            info!("worker_threads: {}", multi);
            runtime::Builder::new_multi_thread()
                .worker_threads(multi)
                .enable_all()
                .thread_name("foobar")
                .build()
                .context("cannot create runtime")?
        }
    };

    let handle = rt.handle();
    let _enter_guard = handle.enter();
    let _ = std::fs::remove_file(SOCK_PATH);
    let listener = UnixListener::bind(SOCK_PATH).unwrap();
    rt.block_on(async move { run_listener(listener).await });
    Ok(())
}

pub async fn run_listener(listener: UnixListener) {
    loop {
        match listener.accept().await {
            Ok((stream, _)) => {
                info!("Received incoming");
                tokio::task::spawn(async move {
                    match handle_client(stream).await {
                        Ok(_) => (),
                        Err(err) => error!("error handling client, error: {}", err),
                    }
                });
            }
            Err(err) => {
                error!("error accepting connection, error: {}", err);
            }
        }
    }
}

async fn handle_client(stream: UnixStream) -> Result<()> {
    let mut buf_stream = BufStream::new(stream);
    let mut line = String::new();
    buf_stream.read_line(&mut line).await?;

    info!("Received request: {}", line);
    buf_stream.write_all(b"END\r\n").await?;
    buf_stream.shutdown().await?;
    drop(buf_stream);
    Ok(())
}

```

## Client code that generates parallel requests

`src/client/main.rs`
```rust
use anyhow::{Context, Result};
use tokio::{
    io::{AsyncWriteExt, BufStream},
    net::UnixStream,
    runtime,
};

pub static SOCK_PATH: &str = "/var/run/sock.sock";

fn main() -> Result<()> {
    let rt = runtime::Builder::new_multi_thread()
        .enable_all()
        .thread_name("resolver-core")
        .build()
        .context("cannot create runtime")?;
    rt.block_on(run_client())?;

    Ok(())
}

async fn run_client() -> Result<()> {
    loop {
        let listener = UnixStream::connect(SOCK_PATH).await?;
        let mut buf_stream = BufStream::new(listener);
        tokio::spawn(async move {
            match buf_stream.write_all(b"foobar\r\n").await {
                Ok(_) => (),
                Err(err) => {
                    println!("write_all error:: {}", err);
                }
            };

            match buf_stream.flush().await {
                Ok(_) => (),
                Err(err) => {
                    println!("flush error:: {}", err);
                }
            };
        });
    }
}

```

## cargo.toml

```toml
[package]
name = "foobar"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[[bin]]
name = "server"
path = "src/main.rs"

[[bin]]
name = "client"
path = "src/client/main.rs"

[dependencies]
tokio = { version = "1.15.0", features = ["full"] }
clap = "3.0.13"
anyhow = "1.0.54"
log = "0.4.17"
env_logger = "0.9.0"

```

I expected to see this happen:

When the server is run as 
```
./server --worker-threads=0
```

and when the client is run as
```
./client
```
I expect to see 

```
[2022-06-21T19:55:57Z ERROR server] error accepting connection, error: Too many open files (os error 24)
```
from the server's stdout once the `ulimit` to openfiles are hit, but I also expect it to recover after the client is stopped.

Instead, this happened: 

With the above code, the server keeps on printing 
```
[2022-06-21T19:55:57Z ERROR server] error accepting connection, error: Too many open files (os error 24)
```
even after the client is exited. I watched the output of `lsof` for some time and it was as below 

```
sudo lsof -p 2704488 | grep -ic sock
1015
```
The number of open file handles to the socket never comes down because it was never released.

### Note: 
This happens only when running in `current_thread`. If the executer is switched to `multi_threaded` by running the server as
`./server --worker-threads=1`, even if the server hits open file limit, it recovers and `lsof` output shows the number of open filehandles to the socket coming down.
I tried to reproduce this in a docker running on my mac, but it didn't occur. I tried running this on baremetal linux and linux running on vmware fusion and I was able to reproduce this.
I have this code added into my repo if anybody want to try it locally on a linux machine. (https://github.com/nohupped/buggy)

### PR DESCRIPTION
**Pull request by Noah-Kennedy**
Fixes #4782.

### CODE REVIEW

**Comment by Darksonn** [comment]
This seems like it might make some operations consume budget twice.

**Comment by Noah-Kennedy** [comment]
Yes, I don't think this is actually where I want to put this.

**Comment by Noah-Kennedy** [comment]
I'll finish this up in the morning, I'm a bit too sleepy to figure out where best to put it right now.

**Comment by Noah-Kennedy** [comment]
@Darksonn I'm not actually sure where something we have could consume budget twice for the same work.

Also, is there any general guidelines for where should consume budget (which operations is it worthwhile for) or not? I was looking through where all we do budgeting vs don't and it doesn't seem super consistent to me.

**Comment by Noah-Kennedy** [comment]
The CI failures I think have to do with tests assuming that something will always immediately yield `Poll::Ready()`, which shouldn't be guaranteed because we reserve the right to put budgeting into things.

**Comment by Darksonn** [comment]
Generally, we consume one budget whenever a poll function returns `Poll::Ready`. However, I agree that this is less clear for readiness-based APIs.

File: tokio/src/io/driver/registration.rs [hunk 1]
```diff
@@ -224,14 +224,21 @@ cfg_io_readiness! {
...
                     return Poll::Ready(Err(io::Error::new(
                         io::ErrorKind::Other,
                         crate::util::error::RUNTIME_SHUTTING_DOWN_ERROR
                     )));
                 }
 
-                Pin::new(&mut fut).poll(cx).map(Ok)
+                let x = Pin::new(&mut fut).poll(cx).map(Ok);
+
+                coop.made_progress();
```
Comment (Darksonn) [inline tokio/src/io/driver/registration.rs:239]:
```suggestion
                if x.is_ready() {
                    coop.made_progress();
                }
```

**Review by Darksonn — COMMENTED** [review]
Looking closer at this, I believe the following should be enough to finish this PR.

**Comment by fasterthanlime** [comment]
@Noah-Kennedy do you have time to finish this up? If not, would you like me to take it over?

**Comment by Noah-Kennedy** [comment]
I should be able to finish this up on Saturday. If you can get to this before then, go for it!

**Comment by Darksonn** [comment]
Since this is a stale draft, I will close it for now.

**Pull request closed without merging**